text/0056-sequence-number-panorama.md

```diff
@@ -0,0 +1,153 @@
+
+
+## Motivation
+
+[motivation]: #motivation
+
+32-byte hashes of earlier units make up the largest part of Highway messages.
+Replacing them by 8-byte sequence numbers will therefore considerably reduce message sizes,
+and improve performance.
+
+
+## Guide-level explanation
+
+[guide-level-explanation]: #guide-level-explanation
+
+Each unit — the main type of message in the Highway protocol — contains a list of hashes of earlier units.
+That way the sender attests to which units they had received before they created the new one.
+This is what defines the directed acyclic graph (DAG) structure of Highway's protocol state:
+The graph has an arrow `u → v` if `u` _directly cites_ `v`, i.e. the message `u` mentions `v` by hash.
+
+What matters to the protocol's logic is not the graph itself but the partial order it defines:
+If `u → v → w` that means the creator of `u` must also have known about `w`, even if `u` doesn't directly cite `w`.
+We write `u > v` if `v` is reachable from `u` by following any number of arrows.
+
+In particular, the actual message `u` that gets sent over the wire does not need to _directly_ cite every `v < u`.
+In the current implementation, `u` contains a _panorama_, which is a list containing for each (honest) validator the hash of the latest `v < u` created by that validator.
+
+We propose replacing the hash with the _sequence number_ instead, i.e. the number of earlier units by the same validator.
+If Alice has created units `a0`, `a1` and `a2` so far, and Bob wants to cite `a2`, instead of its hash, he will now write `2` in the panorama of his unit.
+
+[Panorama]: https://github.com/casper-network/casper-node/blob/b56c29cc97e86ead8d36dfdb2d41a8ecb1dffce9/node/src/components/consensus/highway_core/state/panorama.rs#L25
+
+
+## Reference-level explanation
+
+[reference-level-explanation]: #reference-level-explanation
+
+Units in memory still contain the full hash-based panorama to allow fast lookups.
+
+But before sending a unit `u` over the wire we replace each hash referring to a `v < u` with `v`'s sequence number.
+In addition, we add the hash of `u`'s full hash-based panorama so the recipient can verify after recomputing it.
+And we add `u`'s own predecessor by hash, i.e. the previous unit by the same creator.
+
+The unit wire format now looks like this (the first three fields are new, and the `panorama` field is removed):
+
+```rust
+struct WireUnit {
+    seq_num_panorama: Vec<ObservationSeqNum>,
+    panorama_hash: Hash,
+    previous: Option<Hash>,
+    creator: ValidatorIndex,
+    instance_id: InstanceId,
+    value: Option<BlockPayload>,
+    seq_number: u64,
+    timestamp: Timestamp,
+    round_exp: u8,
+    endorsed: BTreeSet<Hash>,
+}
+
+enum ObservationSeqNum {
+    Correct(u64),
+    Faulty,
+    None,
+}
+```
+
+`None` means that we haven't seen any unit from that validator yet, `Faulty` means they have equivocated.
+`Correct(n)` is the direct citation of the validator's `1 + n`-th unit.
```

Comment:
Why `1+n`-th, and not `n`-th here?

Reply:
Because we start counting at `0`, so the unit with sequence number `n` is the `1 + n`-th unit.

Reply:
Ah, I see! It is slightly confusing, though - maybe something like "the validator's unit with sequence number `n`" would be better, instead?

Reply:
Done in https://github.com/casper-network/ceps/pull/56/commits/d8954d134ac89000dc2a12398e61fdff68ea0663.